plonky2/src/iop/challenger.rs

```diff
@@ -272,6 +285,32 @@ impl<F: RichField + Extendable<D>, H: AlgebraicHasher<F>, const D: usize>
 
         self.input_buffer.clear();
     }
+
+    pub fn duplexing(&mut self, builder: &mut CircuitBuilder<F, D>) {
```

Comment:
If we go with the suggestion above we shouldn't need this. If we do keep it, we could change it to

```rust
if self.input_buffer.is_empty() {
    // Apply the permutation to get a fresh output buffer as from_state expects.
    self.sponge_state = builder.permute::<H>(self.sponge_state);
} else {
    self.absorb_buffered_inputs(builder);
} 
```